llvm/lib/Target/AArch64/AArch64GlobalISelPatterns.td

```diff
@@ -0,0 +1,36 @@
+//===-- AArch64GlobalISelPatterns.td - GlobalISel patterns -*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Selection and combine patterns for GlobalISel.
+//
+//===----------------------------------------------------------------------===//
+
+
+//unpredicate patterns
+
+
+
+// fneg
+def : Pat<(nxv2f64 (fneg nxv2f64:$src)),
+          (FNEG_ZPmZ_D (IMPLICIT_DEF), (PTRUE_D 31), ZPR:$src)>;
```

Comment:
The use of `IMPLICIT_DEF` here can introduce false data-dependcies that will negative impact performance.  We go to some trouble to avoid this in isel (see https://reviews.llvm.org/D105889) whereby when the "real" input is not reused we emit a movprfx to break the dependency.

Reply:
`ag movprfx llvm/lib/Target/AArch64/*.td` movprfx seems to be seldomly used in patterns. Could you give a pointer?

Reply:
I found https://github.com/llvm/llvm-project/pull/116259, but that looks like SVE2.2.

Reply:
`0|(1ULL<<MCID::Pseudo), 0x0ULL },  // Inst #499 = FABS_ZPmZ_H_UNDEF`
pseudo instruction?
from lib/Target/AArch64/AArch64GenInstrInfo.inc

Reply:
```
; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
; RUN: llc -global-isel -mtriple aarch64 -mattr=+sve -global-isel-abort=2 -stop-after=instruction-select -aarch64-enable-gisel-sve=1 %s -o - | FileCheck %s

define <vscale x 8 x half> @fabsnxv8half(<vscale x 8 x half> %a) {
  ; CHECK-LABEL: name: fabsnxv8half
  ; CHECK: bb.1.entry:
  ; CHECK-NEXT:   liveins: $z0
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:zpr = COPY $z0
  ; CHECK-NEXT:   [[PTRUE_H:%[0-9]+]]:ppr_3b = PTRUE_H 31, implicit $vg
  ; CHECK-NEXT:   [[DEF:%[0-9]+]]:zpr = IMPLICIT_DEF
  ; CHECK-NEXT:   [[FABS_ZPmZ_H:%[0-9]+]]:zpr = FABS_ZPmZ_H [[DEF]], [[PTRUE_H]], [[COPY]]
  ; CHECK-NEXT:   $z0 = COPY [[FABS_ZPmZ_H]]
  ; CHECK-NEXT:   RET_ReallyLR implicit $z0
entry:
   %c = tail call <vscale x 8 x half> @llvm.fabs.nxv8f16(<vscale x 8 x half> %a)
  ret <vscale x 8 x half> %c
}
;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
; CHECK: {{.*}}
```
Almost the difference is the undef. It is not in the td pattern.

Reply:
> `0|(1ULL<<MCID::Pseudo), 0x0ULL }, // Inst #499 = FABS_ZPmZ_H_UNDEF` pseudo instruction? from lib/Target/AArch64/AArch64GenInstrInfo.inc

Yes it is pseudo instruction expanded by the code added by https://reviews.llvm.org/D105889 that was shared earlier in this conversation.

Reply:
```
; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
; RUN: llc -global-isel -mtriple aarch64 -mattr=+sve -global-isel-abort=2 -stop-after=instruction-select -aarch64-enable-gisel-sve=1 %s -o - | FileCheck %s

define <vscale x 8 x half> @fabsnxv8half(<vscale x 8 x half> %a) {
  ; CHECK-LABEL: name: fabsnxv8half
  ; CHECK: bb.1.entry:
  ; CHECK-NEXT:   liveins: $z0
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:zpr = COPY $z0
  ; CHECK-NEXT:   [[PTRUE_H:%[0-9]+]]:ppr_3b = PTRUE_H 31, implicit $vg
  ; CHECK-NEXT:   [[DEF:%[0-9]+]]:zpr = IMPLICIT_DEF
  ; CHECK-NEXT:   [[FABS_ZPmZ_H_UNDEF:%[0-9]+]]:zpr = FABS_ZPmZ_H_UNDEF [[DEF]], [[PTRUE_H]], [[COPY]]
  ; CHECK-NEXT:   $z0 = COPY [[FABS_ZPmZ_H_UNDEF]]
  ; CHECK-NEXT:   RET_ReallyLR implicit $z0
entry:
   %c = tail call <vscale x 8 x half> @llvm.fabs.nxv8f16(<vscale x 8 x half> %a)
  ret <vscale x 8 x half> %c
}
;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
; CHECK: {{.*}}
```
I can get the _UNDEF, but I don't get the `killed`.

Reply:
I'm not concerned about the kill flag.